### PR TITLE
Additional Incremental Build Tests

### PR DESCRIPTION
**Pull request by TheOtherDave**

### CODE REVIEW

**Comment by TheOtherDave** [comment]
IIRC, all of the places where I'm not sure I'm doing the right thing are marked with a FIXME

File: Fixtures/IncrementalBuildTests/SourceCodeError/Sources/main.swift [hunk 1]
```diff
@@ -0,0 +1,2 @@
+print(firstLine)
+print(secondLine)
```
Comment (ddunbar) [inline Fixtures/IncrementalBuildTests/SourceCodeError/Sources/main.swift]: Is this test package any different from `PackageError`? I couldn't spot the different but its kind hard to see in the PR review view...

File: Tests/BuildTests/IncrementalBuildTests.swift [hunk 2]
```diff
@@ -15,7 +15,130 @@ import Basic
...
+            }
+            
+            // So the idea here is to log everything in ".builds/debug/", except
+            // for "ModuleCache", because that doesn't seem relevant. We crash
+            // without the `&& !$0.hasSuffix(".dSYM")` part, probably because
+            // localFileSystem.isDirectory() seems to return true for .dSYM files?
+            var bp = try! localFileSystem.getDirectoryContents(path.appending(components: ".build", "debug"))
+                .filter { $0 != "ModuleCache" }
+            // Get the list of "files" that are actually directories...
+            var bpd = bp.filter { localFileSystem.isDirectory(path.appending(components: ".build", "debug", $0)) && !$0.hasSuffix(".dSYM") }
```
Comment (ddunbar) [inline Tests/BuildTests/IncrementalBuildTests.swift]: I believe it should be possible to simplify this by using Basic.walk (see `PathShims.swift`)...

File: Tests/BuildTests/IncrementalBuildTests.swift [hunk 3]
```diff
@@ -15,7 +15,130 @@ import Basic
...
+@discardableResult
+func checkForNils(_ logs: BuildLog, allowableFailures: [Builds] = []) -> [Builds] {
+    return logs.keys.filter {
+        (logs[$0] == nil && !($0 == Builds.null && !alwaysDoRedundancyCheck)) || (logs[$0]!.logWasNil && !allowableFailures.contains($0))
+    }
+}
+
+typealias BuildLog = [Builds : BuildResult]
 
+// Seems like this is too tied to the build's log format
```
Comment (ddunbar) [inline Tests/BuildTests/IncrementalBuildTests.swift]: Agreed, we should eventually have a mode for llbuild that just dumps a streaming JSON log so it can be parsed mechanically.

File: Tests/BuildTests/IncrementalBuildTests.swift [hunk 4]
```diff
@@ -54,22 +181,293 @@ final class IncrementalBuildTests: XCTestCase {
...
+            let extraFile = BufferedOutputByteStream()
+            
+            // We want to be careful with the filesystem here. There's not much
+            // we can do to recover from an error writing the file back out, but
+            // if we get an error reading the file in in the first place,
+            // something has gone really wrong.
+            do {
+                extraFile <<< (try localFileSystem.readFileContents(extraFilePath))
+            } catch let error {
+                let message = "\(line())caught \"\(error)\" trying to read \(extraFilePath.asString). Aborting test"
```
Comment (ddunbar) [inline Tests/BuildTests/IncrementalBuildTests.swift]: Any reason this test is careful but not the others? One trick that I use is to just mark the test case itself as "throws" and let the error throw all the way out of the test, if isn't the kind of error I am actually writing the test to normally catch with a richer diagnostic.

File: Fixtures/IncrementalBuildTests/SourceCodeError/Sources/main.swift (hunk 1 above)
Comment (TheOtherDave) [inline Fixtures/IncrementalBuildTests/SourceCodeError/Sources/main.swift]: They're similar, but not identical. This test makes a syntax error in a source code file ("SecondLine.swift"), whereas the PackageError test makes a syntax error in "Package.swift".

File: Tests/BuildTests/IncrementalBuildTests.swift (hunk 4 above)
Comment (TheOtherDave) [inline Tests/BuildTests/IncrementalBuildTests.swift]: Only that explicitly deleting a file without the user requesting it kinda makes me itch. OTOH, I suppose you could argue that the user _did_ request it by running the test :-)
You're right, though... The extra checks here should probably either be in all the tests or none of them.

File: Tests/BuildTests/IncrementalBuildTests.swift (hunk 2 above)
Comment (TheOtherDave) [inline Tests/BuildTests/IncrementalBuildTests.swift]: Oh, hey! Yep, that seems easier.

**Comment by aciidb0mb3r** [comment]
Closing this due to age but feel free to reopen.
